plugins/inputs/monit/monit.go

```diff
@@ -0,0 +1,434 @@
+	Memory           Memory  `xml:"memory"`
+	CPU              CPU     `xml:"cpu"`
+	System           System  `xml:"system"`
+	Size             int64   `xml:"size"`
+	Mode             int     `xml:"mode"`
+	Program          Program `xml:"program"`
+	Block            Block   `xml:"block"`
+	Inode            Inode   `xml:"inode"`
+	Pid              int64   `xml:"pid"`
+	ParentPid        int64   `xml:"ppid"`
+	Threads          int     `xml:"threads"`
+	Children         int     `xml:"children"`
+	Port             Port    `xml:"port"`
+	Link             Link    `xml:"link"`
+}
+
+type Link struct {
+	State    int      `xml:"state"`
+	Speed    int64    `xml:"speed"`
+	Duplex   int      `xml:"duplex"`
+	Download Download `xml:"download"`
+	Upload   Upload   `xml:"upload"`
+}
+
+type Download struct {
+	Packets struct {
+		Now   int64 `xml:"now"`
+		Total int64 `xml:"total"`
+	} `xml:"packets"`
+	Bytes struct {
+		Now   int64 `xml:"now"`
+		Total int64 `xml:"total"`
+	} `xml:"bytes"`
+	Errors struct {
+		Now   int64 `xml:"now"`
+		Total int64 `xml:"total"`
+	} `xml:"errors"`
+}
+
+type Upload struct {
+	Packets struct {
+		Now   int64 `xml:"now"`
+		Total int64 `xml:"total"`
+	} `xml:"packets"`
+	Bytes struct {
+		Now   int64 `xml:"now"`
+		Total int64 `xml:"total"`
+	} `xml:"bytes"`
+	Errors struct {
+		Now   int64 `xml:"now"`
+		Total int64 `xml:"total"`
+	} `xml:"errors"`
+}
+
+type Port struct {
+	Hostname   string `xml:"hostname"`
+	PortNumber int64  `xml:"portnumber"`
+	Request    string `xml:"request"`
+	Protocol   string `xml:"protocol"`
+	Type       string `xml:"type"`
+}
+
+type Block struct {
+	Percent float64 `xml:"percent"`
+	Usage   float64 `xml:"usage"`
+	Total   float64 `xml:"total"`
+}
+
+type Inode struct {
+	Percent float64 `xml:"percent"`
+	Usage   float64 `xml:"usage"`
+	Total   float64 `xml:"total"`
+}
+
+type Program struct {
+	Started int64 `xml:"started"`
+	Status  int   `xml:"status"`
+}
+
+type Memory struct {
+	Percent       float64 `xml:"percent"`
+	PercentTotal  float64 `xml:"percenttotal"`
+	Kilobyte      int64   `xml:"kilobyte"`
+	KilobyteTotal int64   `xml:"kilobytetotal"`
+}
+
+type CPU struct {
+	Percent      float64 `xml:"percent"`
+	PercentTotal float64 `xml:"percenttotal"`
+}
+
+type System struct {
+	Load struct {
+		Avg01 float64 `xml:"avg01"`
+		Avg05 float64 `xml:"avg05"`
+		Avg15 float64 `xml:"avg15"`
+	} `xml:"load"`
+	CPU struct {
+		User   float64 `xml:"user"`
+		System float64 `xml:"system"`
+		Wait   float64 `xml:"wait"`
+	} `xml:"cpu"`
+	Memory struct {
+		Percent  float64 `xml:"percent"`
+		Kilobyte int64   `xml:"kilobyte"`
+	} `xml:"memory"`
+	Swap struct {
+		Percent  float64 `xml:"percent"`
+		Kilobyte float64 `xml:"kilobyte"`
+	} `xml:"swap"`
+}
+
+type Monit struct {
+	Address  string `toml:"address"`
+	Username string `toml:"username"`
+	Password string `toml:"password"`
+	client   HTTPClient
+	tls.ClientConfig
+	Timeout internal.Duration `toml:"timeout"`
+}
+
+type HTTPClient interface {
```

Comment:
I like that you mock http responses, but there's a simpler way to do it that is built in to http.Client.  Before calling your http.Client's .Do(), you set its Transport to a custom http.RoundTripper.

I'd prefer to remove this HTTPClient interface and use a custom RoundTripper because it requires less code and because it's standard Go so it will be easier for whoever works on this code later.

https://golang.org/pkg/net/http/#RoundTripper

Reply:
Thank you for the comments Reimda. I have written some sample code around RoundTripper. Could you please check?

type transportMock struct {
        errorString string
}

func newTransportMock(errorString string) http.RoundTripper {
        return &transportMock{
                errorString:    errorString,
        }
}

func (t *transportMock) RoundTrip(r *http.Request) (*http.Response, error) {
        return nil, errors.New(t.errorString)
}

func TestAllowHosts(t *testing.T) {

        networkError := "Get http://127.0.0.1:2812/_status?format=xml: " +
                        "read tcp 192.168.10.2:55610->127.0.0.1:2812: " +
                        "read: connection reset by peer"
        r := &Monit{
                Address:  "http://127.0.0.1:2812",
                Username: "test",
                Password: "test",
                client:         &http.Client{}
        }

        var acc testutil.Accumulator

        r.client.Transport = newTransportMock(networkError)

        err := r.Gather(&acc)

        if assert.Error(t, err) {
                assert.Contains(t, err.Error(), "read: connection reset by peer")
        }
}

Reply:
That's the idea.  Also change Monit.client to be a http.Client, not a *http.Client, so you don't have to explicitly initialize it.  I'd probably make type transportMock an empty struct and put the networkError string right in the RoundTrip func.  Then you don't need newTransportMock either.

Looks like we already merged this PR.  If you want to make a new PR to change this, feel free to add me as a reviewer.

thanks

Reply:
Thank you for the comments. I have raised a new PR. Could you please review?
https://github.com/influxdata/telegraf/pull/6955